README
CGI::Application::Server version 0.062
======================================

See the individual module documentation for more information

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

    HTTP::Request
    HTTP::Status
    CGI::Application
    HTTP::Server::Simple
    HTTP::Server::Simple::Static

And requires these modules for testing:

    Test::More
    Test::Exception
    Test::HTTP::Server::Simple
    Test::WWW::Mechanize
    CGI::Application::Plugin::Redirect

COPYRIGHT AND LICENCE

Copyright (C) 2006 Infinity Interactive, Inc.

http://www.iinteractive.com

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 



Changes
Changes for CGI::Application::Server

{{$NEXT}}
          point bugtracker back at RT; rjbs is abandoning this library

0.063     2013-07-05 21:08:53 America/New_York
          repackage, update bugtracker

0.062			2010-01-20
					work on alternate ports (thanks, Potatohead @ github)
					misc. cleanup (Potatohead, rjbs)

0.061			2009-02-08
					reinitialize $query between runs when entry point is an object
					(patch from JALDHAR, rt #42199)

0.060 Sun. Nov  2, 2008
          a number of patches from George Hartzell; thanks!
          documenation improvements
          use HTTP::Server::Simple::Static as its documentation suggests
          add handling for a "/" entry point

0.050 Sun. Oct 21, 2007
          add support for CGI::Application objects as entry points
          handle PATH_INFO munging for CGI::App, not just ::Dispatch
          both these changes thanks to JALDHAR

0.04 Sun. Oct 21, 2007
          add support for CGI::Application::Dispatch classes

0.03 Tues. Feb 20, 2007
          Fixed is_valid_entry_point to be more strict
          in how it matched URIs, added tests

          fixed the way the captured stdout is recombined
          to make steaming non-HTML mime-types work

0.02 Mon. April 10, 2006
          Added logging of every request (uri, entry-point, and params)

0.01 Wed. March 15, 2006
          initial release of module


t/htdocs/test.js

function greet () { 
	alert("Hello World");
}

t/lib/MyCGIApp.pm
package MyCGIApp;

use base 'CGI::Application';

use CGI::Application::Plugin::Redirect;

sub setup {
	my $self = shift;
	$self->start_mode('mode1');
	$self->mode_param('rm');
	$self->run_modes(
	        'mode1' => 'hello_world',
	        'mode2' => 'goodbye_world',		
            'mode3' => 'redirected',
            'mode4' => 'redirect_end',
	);
}	

sub hello_world {
	return "<HTML><TITLE>Hello</TITLE><BODY><H1>Hello World!</H1><HR>" . 
		   "<A HREF='index.cgi?rm=mode2'>Goodbye</A>" . 
		   "<A HREF='index.cgi?rm=mode3'>Redirected</A>" . 
		   "</BODY></HTML>";
}

sub goodbye_world {
	return "<HTML><TITLE>Goodbye</TITLE><BODY><H1>Goodbye World!</H1><HR>" . 
	       "<A HREF='index.cgi?rm=mode1'>Hello</A>" . 
	   	   "</BODY></HTML>";		
}

sub redirected {
    my $self = shift;
    return $self->redirect( "index.cgi?rm=mode4" );
}

sub redirect_end {
	return "<HTML><TITLE>Redirect End</TITLE><BODY><H1>Redirected!</H1><HR>" . 
	       "<A HREF='index.cgi?rm=mode1'>Back to Hello</A>" . 
	   	   "</BODY></HTML>";		
}

1;


lib/CGI/Application/Server.pm
use strict;
use warnings;
package CGI::Application::Server;
# ABSTRACT: a simple HTTP server for developing with CGI::Application

use Carp 0.01 qw( confess );
use CGI qw( param );
use Scalar::Util 1.18 qw( blessed reftype );
use HTTP::Response;
use HTTP::Status;

use base qw( HTTP::Server::Simple::CGI );
use HTTP::Server::Simple 0.18;
use HTTP::Server::Simple::Static 0.02;

# HTTP::Server::Simple methods

sub new {
    my $class = shift;
    my $self  = $class->SUPER::new(@_);
    $self->{entry_points} = {};
    $self->{document_root}  = '.';
    return $self;
}

# accessors

sub document_root {
    my ($self, $document_root) = @_;
    if (defined $document_root) {
        (-d $document_root)
            || confess "The server root ($document_root) is not found";
        $self->{document_root} = $document_root;
    }
    $self->{document_root};
}

sub entry_points {
    my ($self, $entry_points) = @_;
    if (defined $entry_points) {
        (reftype($entry_points) && reftype($entry_points) eq 'HASH')
            || confess "The entry points map must be a HASH reference, not $entry_points";
        $self->{entry_points} = $entry_points;
    }
    $self->{entry_points};
}

# check request

sub is_valid_entry_point {
    my ($self, $uri) = @_;

    # Remove all parameters
    $uri =~ s/\?.*//;

    while ( $uri ) {
        # Check to see if this is an exact match
        if (exists $self->{entry_points}{$uri}) {
            return ($uri, $self->{entry_points}{$uri});
        }

        # Remove the rightmost path element
        $uri =~ s/\/[^\/]*$//;
    }

    # Check to see if there's an entry for '/'
    if (exists $self->{entry_points}{'/'}) {
	return ($uri, $self->{entry_points}{'/'});
    }

    # Didn't find anything. Oh, well.
    return;
}

sub handle_request {
    my ($self, $cgi) = @_;
    if (my ($path, $target) = $self->is_valid_entry_point($ENV{REQUEST_URI})) {
        # warn "$ENV{REQUEST_URI} ($target)\n";
        # warn "\t$_ => " . param( $_ ) . "\n" for param();

        local $ENV{CGI_APP_RETURN_ONLY} = 1;
        (local $ENV{PATH_INFO} = $ENV{PATH_INFO}) =~ s/\A\Q$path//;

        if (-d $target && -x $target) {
	  return $self->serve_static($cgi, $target);
	}
	elsif ($target->isa('CGI::Application::Dispatch')) {
	  return $self->_serve_response($target->dispatch);
        } elsif ($target->isa('CGI::Application')) {
          if (!defined blessed $target) {
	    return $self->_serve_response($target->new->run);
          } else {
        $target->query($cgi);
	    return $self->_serve_response($target->run);
          }
	}
	else {
          confess "Target must be a CGI::Application or CGI::Application::Dispatch subclass or the name of a directory that exists and is readable.\n";
        }
    } else {
        return $self->serve_static($cgi, $self->document_root);
    }
}

sub _serve_response {
  my ( $self, $stdout ) = @_;

  my $response = $self->_build_response( $stdout );
  print $response->as_string();

  return 1;			# Like ...Simple::Static::serve_static does
}

# Shamelessly stolen from HTTP::Request::AsCGI by chansen
sub _build_response {
    my ( $self, $stdout ) = @_;

    $stdout =~ s{(.*?\x0d?\x0a\x0d?\x0a)}{}xsm;
    my $headers = $1;

    unless ( defined $headers ) {
        $headers = "HTTP/1.1 500 Internal Server Error\x0d\x0a";
    }

    unless ( $headers =~ /^HTTP/ ) {
        $headers = "HTTP/1.1 200 OK\x0d\x0a" . $headers;
    }

    my $response = HTTP::Response->parse($headers);
    $response->date( time() ) unless $response->date;

    my $message = $response->message;
    my $status  = $response->header('Status');

    $response->header( Connection => 'close' );

    if ( $message && $message =~ /^(.+)\x0d$/ ) {
        $response->message($1);
    }

    if ( $status && $status =~ /^(\d\d\d)\s?(.+)?$/ ) {

        my $code    = $1;
        $message = $2 || HTTP::Status::status_message($code);

        $response->code($code);
        $response->message($message);
    }

    my $length = length $stdout;

    if ( $response->code == 500 && !$length ) {

        $response->content( $response->error_as_HTML );
        $response->content_type('text/html');

        return $response;
    }

    $response->add_content($stdout);
    $response->content_length($length);

    return $response;
}


1;

__END__

=head1 SYNOPSIS

  use CGI::Application::Server;
  use MyCGIApp;
  use MyCGIApp::Admin;
  use MyCGI::App::Account::Dispatch;
  use MyCGIApp::DefaultApp;

  my $server = CGI::Application::Server->new();

  # this CGI::Application object will stay persistent, might not be safe to use
  # in this way - your mileage may vary
  # http://www.mail-archive.com/cgiapp@lists.erlbaum.net/msg08997.html
  my $object = MyOtherCGIApp->new(PARAMS => { foo => 1, bar => 2 });

  $server->document_root('./htdocs');
  $server->entry_points({
      '/'          => 'MyCGIApp::DefaultApp',
      '/index.cgi' => 'MyCGIApp',
      '/admin'     => 'MyCGIApp::Admin',
      '/account'   => 'MyCGIApp::Account::Dispatch',
      '/users'     => $object,
      '/static'    => '/usr/local/htdocs',
  });
  $server->run();

=head1 DESCRIPTION

This is a simple HTTP server for for use during development with
L<CGI::Application>. At this moment, it serves our needs in a
very basic way. The plan is to release early and release often,
and add features when we need them. That said, we welcome any
and all patches, tests and feature requests (the ones with which
are accompanied by failing tests will get priority).

=method B<new ($port)>

This acts just like C<new> for L<HTTP::Server::Simple>, except it
will initialize instance slots that we use.

=method B<handle_request>

This will check the request uri and dispatch appropriately, either
to an entry point, or serve a static file (html, jpeg, gif, etc).

=method B<entry_points (?$entry_points)>

This accepts a HASH reference in C<$entry_points>, which maps server entry
points (uri) to L<CGI::Application> or L<CGI::Application::Dispatch> class
names or objects or to directories from which static content will be served
by HTTP::Server::Simple::Static.  See the L<SYNOPSIS> above for examples.

=method B<is_valid_entry_point ($uri)>

This attempts to match the C<$uri> to an entry point.

=method B<document_root (?$document_root)>

This is the server's document root where all static files will
be served from.

=head1 CAVEATS

This is a subclass of L<HTTP::Server::Simple> and all of its caveats
apply here as well.

=head1 ACKNOWLEDGEMENTS

The HTTP response handling was shamelessly stolen from L<HTTP::Request::AsCGI>
by chansen

=cut


lib/HTTP/Server/Simple/CGI/Application.pm

package HTTP::Server::Simple::CGI::Application;

use strict;
use warnings;
use Carp;

use Carp         'confess';
use Scalar::Util 'blessed', 'reftype';

our $VERSION = '0.01';

use base qw(HTTP::Server::Simple::CGI HTTP::Server::Simple::Static);

# HTTP::Server::Simple methods

sub new {
	my $class = shift;
	my $self  = $class->SUPER::new(@_); 
	$self->{entry_points} = {};	
	$self->{server_root}  = '.';
	return $self;
}

# accessors

sub server_root {
	my ($self, $server_root) = @_;
	if (defined $server_root) {
		(-d $server_root)
			|| confess "The server root ($server_root) is not found";
		$self->{server_root} = $server_root;
	}
	$self->{server_root};
}

sub entry_points {
	my ($self, $entry_points) = @_;
	if (defined $entry_points) {
		(reftype($entry_points) && reftype($entry_points) eq 'HASH')
			|| confess "The entry points map must be a HASH reference, not $entry_points";
		$self->{entry_points} = $entry_points;
	}
	$self->{entry_points};	
}

# check request

sub is_valid_entry_point {
	my ($self, $uri) = @_;
	foreach my $entry_point (keys %{$self->{entry_points}}) {
		return $self->{entry_points}->{$entry_point}
			if index($uri, $entry_point) == 0;
	}
	return undef;
}

sub handle_request {
	my ($self, $cgi) = @_;
	if (my $entry_point = $self->is_valid_entry_point($ENV{REQUEST_URI})) {
		# NOTE:
		# this does not handle Redirects correctly,.. 
		# how should we handle those?
    	print "HTTP/1.0 200 OK\r\n";
		$entry_point->new->run;		
	}
	else {
    	$self->serve_static($cgi, $self->server_root);
	} 
}

1;

__END__

=pod

=head1 NAME

HTTP::Server::Simple::CGI::Application - A HTTP::Server::Simple subclass for developing CGI::Application

=head1 SYNOPSIS

  use HTTP::Server::Simple::CGI::Application;

  my $server = HTTP::Server::Simple::CGI::Application->new();
  $server->server_root('./htdocs');
  $server->entry_points({
	  '/index.cgi' => 'MyCGIApp',
	  '/admin'     => 'MyCGIApp::Admin'
  });
  $server->run();

=head1 DESCRIPTION

This is a simple L<HTTP::Server::Simple> subclass for use during 
development with L<CGI::Appliaction>. 

=head1 METHODS

=head2 HTTP::Server::Simple methods

These are overridden L<HTTP::Server::Simple> methods.

=over 4

=item B<new ($port)>

This acts just like C<new> for L<HTTP::Server::Simple>, except it 
will initialize instance slots that we use.

=item B<handle_request>

This will check the request uri and dispatch appropriately.

=back

=head2 Accessors & Utils

=over 4

=item B<entry_points (?$entry_points)>

=item B<is_valid_entry_point ($uri)>

This performs the entry point to C<$uri> match test.

=item B<server_root (?$server_root)>

This is the server's document root where all static files will 
be served from.

=back

=head1 BUGS

All complex software has bugs lurking in it, and this module is no 
exception. If you find a bug please either email me, or add the bug
to cpan-RT.

=head1 CODE COVERAGE

I use L<Devel::Cover> to test the code coverage of my tests, below 
is the L<Devel::Cover> report on this module's test suite.

=head1 ACKNOWLEDGEMENTS

=head1 AUTHOR

Stevan Little E<lt>stevan@iinteractive.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright 2006 by Infinity Interactive, Inc.

L<http://www.iinteractive.com>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 

=cut


Build.PL
use Module::Build;

use strict;

my $build = Module::Build->new(
    module_name => 'CGI::Application::Server',
    license => 'perl',
    requires => {
        'Scalar::Util'  => '1.18',
        'Carp'          => '0.01',
        'HTTP::Request' => '0',
        'HTTP::Status'  => '0',
        'CGI::Application'             => '4.21',
        'HTTP::Server::Simple'         => '0.18',
        'HTTP::Server::Simple::Static' => '0.02',
    },
    optional => {
    },
    build_requires => {
        'Test::More'                 => '0.47',
        'Test::Exception'            => '0.21',
        'Test::HTTP::Server::Simple' => '0.02',
        'Test::WWW::Mechanize'       => '1.08',
        'CGI::Application::Plugin::Redirect' => '0',
    },
    recursive_test_files => 1,
    add_to_cleanup => [
        'META.yml', '*.bak', '*.gz',
    ],
);

$build->create_build_script;

